// @diplodoc/algolia-extension/dist/core/document-processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractHeadings = extractHeadings;
exports.splitAndAddLargeRecord = splitAndAddLargeRecord;
exports.getRecordSize = getRecordSize;
exports.splitDocumentIntoSections = splitDocumentIntoSections;
exports.processDocument = processDocument;
const cheerio_1 = require("cheerio");
const indexer_1 = require("@diplodoc/search-extension/indexer");
const utils_1 = require("@diplodoc/cli/lib/utils");
function extractHeadings(html) {
    const $ = (0, cheerio_1.load)(html);
    const headings = [];
    $('h1, h2, h3, h4, h5, h6').each((_, element) => {
        const textPieces = $(element)
            .contents()
            .map((_, el) => $(el).text())
            .get();
        const uniqueText = [...new Set(textPieces)].join('').trim();
        if (uniqueText) {
            headings.push(uniqueText);
        }
    });
    return headings;
}
function splitAndAddLargeRecord(record, acc, chunkSize = 4000) {
    const baseObjectID = record.objectID;
    const content = record.content;
    const chunks = Math.ceil(content.length / chunkSize);
    if (chunks <= 1) {
        acc.push(record);
        return;
    }
    for (let i = 0; i < chunks; i++) {
        const start = i * chunkSize;
        const end = Math.min(start + chunkSize, content.length);
        const chunkContent = content.slice(start, end);
        const chunkRecord = {
            ...record,
            objectID: `${baseObjectID}-chunk-${i + 1}`,
            content: chunkContent,
        };
        acc.push(chunkRecord);
    }
}
function getRecordSize(record) {
    const jsonString = JSON.stringify(record);
    return Buffer.from(jsonString).length;
}
function splitDocumentIntoSections(html) {
    const $ = (0, cheerio_1.load)(html);
    const sections = [];
    let currentSection = { heading: '', content: '', anchor: '' };
    let mainHeading = '';
    $('body')
        .children()
        .each((_, element) => {
        const $el = $(element);
        if ($el.is('h1, h2, h3, h4, h5, h6')) {
            if (currentSection.content.trim()) {
                sections.push({ ...currentSection });
            }
            const level = $el[0].name[1];
            let headingText = '';
            for (const el of $el.contents()) {
                if (el.type === 'text') {
                    headingText = String(el.data);
                }
            }
            if (!headingText) {
                headingText = $el.text().trim();
            }
            if (!mainHeading) {
                mainHeading = headingText;
            }
            currentSection = {
                heading: headingText,
                anchor: $el.attr('id') || '',
                content: '',
                level: parseInt(level, 10),
            };
        }
        else {
            currentSection.content += $el.text().trim() + ' ';
        }
    });
    if (currentSection.content.trim()) {
        sections.push({ ...currentSection });
    }
    return { sections, mainHeading };
}
function createBaseRecord(path, lang, title, keywords = [], skipHtmlExtension) {
    const url = path.replace(/\.\w+$/, '') + '.html';
    const prettyUrl = skipHtmlExtension ? (0, utils_1.shortLink)(url) : url;
    return {
        title,
        keywords,
        url: prettyUrl,
        lang,
    };
}
function processDocument(context) {
    const { path, lang, html, title, meta, skipHtmlExtension } = context;
    const { sections, mainHeading } = splitDocumentIntoSections(html);
    const records = [];
    const baseTitle = title || meta.title || mainHeading || '';
    const baseKeywords = meta.keywords || [];
    const MAX_RECORD_SIZE = 9600;
    if (sections.length === 0) {
        const baseRecord = createBaseRecord(path, lang, baseTitle, baseKeywords, skipHtmlExtension);
        const record = {
            ...baseRecord,
            objectID: path.replace(/\.\w+$/, ''),
            content: (0, indexer_1.html2text)(html).slice(0, 5000),
            headings: extractHeadings(html),
            anchor: '',
            level: 1,
        };
        if (getRecordSize(record) >= MAX_RECORD_SIZE) {
            splitAndAddLargeRecord(record, records);
        }
        else {
            records.push(record);
        }
        return records;
    }
    sections.forEach((section, index) => {
        const baseRecord = createBaseRecord(path, lang, baseTitle, baseKeywords, skipHtmlExtension);
        const record = {
            ...baseRecord,
            objectID: `${path.replace(/\.\w+$/, '')}-${index}`,
            content: section.content.trim(),
            headings: [section.heading],
            anchor: section.anchor,
            section: section.heading || undefined,
            level: section.level || 1,
        };
        if (getRecordSize(record) >= MAX_RECORD_SIZE) {
            splitAndAddLargeRecord(record, records);
        }
        else {
            records.push(record);
        }
    });
    return records;
}
