// @diplodoc/search-extension/lib/indexer/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/indexer/index.ts
var indexer_exports = {};
__export(indexer_exports, {
  Indexer: () => Indexer,
  ReleaseFormat: () => ReleaseFormat,
  html2text: () => html2text
});
module.exports = __toCommonJS(indexer_exports);
var import_lunr59 = __toESM(require("lunr"));

// src/constants.ts
var INDEX_FIELDS = {
  title: 5,
  keywords: 4,
  content: 3
};

// src/indexer/langs/ar.ts
var import_lunr_stemmer = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr2 = __toESM(require("lunr-languages/lunr.ar"));
function ar(lunr2) {
  (0, import_lunr_stemmer.default)(lunr2);
  (0, import_lunr2.default)(lunr2);
  (0, import_lunr.default)(lunr2);
  return lunr2.ar;
}

// src/indexer/langs/da.ts
var import_lunr_stemmer2 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr3 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr4 = __toESM(require("lunr-languages/lunr.da"));
function da(lunr2) {
  (0, import_lunr_stemmer2.default)(lunr2);
  (0, import_lunr4.default)(lunr2);
  (0, import_lunr3.default)(lunr2);
  return lunr2.da;
}

// src/indexer/langs/de.ts
var import_lunr_stemmer3 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr5 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr6 = __toESM(require("lunr-languages/lunr.de"));
function de(lunr2) {
  (0, import_lunr_stemmer3.default)(lunr2);
  (0, import_lunr6.default)(lunr2);
  (0, import_lunr5.default)(lunr2);
  return lunr2.de;
}

// src/indexer/langs/du.ts
var import_lunr_stemmer4 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr7 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr8 = __toESM(require("lunr-languages/lunr.du"));
function du(lunr2) {
  (0, import_lunr_stemmer4.default)(lunr2);
  (0, import_lunr8.default)(lunr2);
  (0, import_lunr7.default)(lunr2);
  return lunr2.du;
}

// src/indexer/langs/el.ts
var import_lunr_stemmer5 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr9 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr10 = __toESM(require("lunr-languages/lunr.el"));
function el(lunr2) {
  (0, import_lunr_stemmer5.default)(lunr2);
  (0, import_lunr10.default)(lunr2);
  (0, import_lunr9.default)(lunr2);
  return lunr2.el;
}

// src/indexer/langs/es.ts
var import_lunr_stemmer6 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr11 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr12 = __toESM(require("lunr-languages/lunr.es"));
function es(lunr2) {
  (0, import_lunr_stemmer6.default)(lunr2);
  (0, import_lunr12.default)(lunr2);
  (0, import_lunr11.default)(lunr2);
  return lunr2.es;
}

// src/indexer/langs/fi.ts
var import_lunr_stemmer7 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr13 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr14 = __toESM(require("lunr-languages/lunr.fi"));
function fi(lunr2) {
  (0, import_lunr_stemmer7.default)(lunr2);
  (0, import_lunr14.default)(lunr2);
  (0, import_lunr13.default)(lunr2);
  return lunr2.fi;
}

// src/indexer/langs/fr.ts
var import_lunr_stemmer8 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr15 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr16 = __toESM(require("lunr-languages/lunr.fr"));
function fr(lunr2) {
  (0, import_lunr_stemmer8.default)(lunr2);
  (0, import_lunr16.default)(lunr2);
  (0, import_lunr15.default)(lunr2);
  return lunr2.fr;
}

// src/indexer/langs/he.ts
var import_lunr_stemmer9 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr17 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr18 = __toESM(require("lunr-languages/lunr.he"));
function he(lunr2) {
  (0, import_lunr_stemmer9.default)(lunr2);
  (0, import_lunr18.default)(lunr2);
  (0, import_lunr17.default)(lunr2);
  return lunr2.he;
}

// src/indexer/langs/hu.ts
var import_lunr_stemmer10 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr19 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr20 = __toESM(require("lunr-languages/lunr.hu"));
function hu(lunr2) {
  (0, import_lunr_stemmer10.default)(lunr2);
  (0, import_lunr20.default)(lunr2);
  (0, import_lunr19.default)(lunr2);
  return lunr2.hu;
}

// src/indexer/langs/hy.ts
var import_lunr_stemmer11 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr21 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr22 = __toESM(require("lunr-languages/lunr.hy"));
function hy(lunr2) {
  (0, import_lunr_stemmer11.default)(lunr2);
  (0, import_lunr22.default)(lunr2);
  (0, import_lunr21.default)(lunr2);
  return lunr2.hy;
}

// src/indexer/langs/it.ts
var import_lunr_stemmer12 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr23 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr24 = __toESM(require("lunr-languages/lunr.it"));
function it(lunr2) {
  (0, import_lunr_stemmer12.default)(lunr2);
  (0, import_lunr24.default)(lunr2);
  (0, import_lunr23.default)(lunr2);
  return lunr2.it;
}

// src/indexer/langs/ko.ts
var import_lunr_stemmer13 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr25 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr26 = __toESM(require("lunr-languages/lunr.ko"));
function ko(lunr2) {
  (0, import_lunr_stemmer13.default)(lunr2);
  (0, import_lunr26.default)(lunr2);
  (0, import_lunr25.default)(lunr2);
  return lunr2.ko;
}

// src/indexer/langs/nl.ts
var import_lunr_stemmer14 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr27 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr28 = __toESM(require("lunr-languages/lunr.nl"));
function nl(lunr2) {
  (0, import_lunr_stemmer14.default)(lunr2);
  (0, import_lunr28.default)(lunr2);
  (0, import_lunr27.default)(lunr2);
  return lunr2.nl;
}

// src/indexer/langs/no.ts
var import_lunr_stemmer15 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr29 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr30 = __toESM(require("lunr-languages/lunr.no"));
function no(lunr2) {
  (0, import_lunr_stemmer15.default)(lunr2);
  (0, import_lunr30.default)(lunr2);
  (0, import_lunr29.default)(lunr2);
  return lunr2.no;
}

// src/indexer/langs/pt.ts
var import_lunr_stemmer16 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr31 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr32 = __toESM(require("lunr-languages/lunr.pt"));
function pt(lunr2) {
  (0, import_lunr_stemmer16.default)(lunr2);
  (0, import_lunr32.default)(lunr2);
  (0, import_lunr31.default)(lunr2);
  return lunr2.pt;
}

// src/indexer/langs/ro.ts
var import_lunr_stemmer17 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr33 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr34 = __toESM(require("lunr-languages/lunr.ro"));
function ro(lunr2) {
  (0, import_lunr_stemmer17.default)(lunr2);
  (0, import_lunr34.default)(lunr2);
  (0, import_lunr33.default)(lunr2);
  return lunr2.ro;
}

// src/indexer/langs/ru.ts
var import_lunr_stemmer18 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr35 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr36 = __toESM(require("lunr-languages/lunr.ru"));
function ru(lunr2) {
  (0, import_lunr_stemmer18.default)(lunr2);
  (0, import_lunr36.default)(lunr2);
  (0, import_lunr35.default)(lunr2);
  return lunr2.ru;
}

// src/indexer/langs/sv.ts
var import_lunr_stemmer19 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr37 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr38 = __toESM(require("lunr-languages/lunr.sv"));
function sv(lunr2) {
  (0, import_lunr_stemmer19.default)(lunr2);
  (0, import_lunr38.default)(lunr2);
  (0, import_lunr37.default)(lunr2);
  return lunr2.sv;
}

// src/indexer/langs/tr.ts
var import_lunr_stemmer20 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr39 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr40 = __toESM(require("lunr-languages/lunr.tr"));
function tr(lunr2) {
  (0, import_lunr_stemmer20.default)(lunr2);
  (0, import_lunr40.default)(lunr2);
  (0, import_lunr39.default)(lunr2);
  return lunr2.tr;
}

// src/indexer/langs/vi.ts
var import_lunr_stemmer21 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr41 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr42 = __toESM(require("lunr-languages/lunr.vi"));
function vi(lunr2) {
  (0, import_lunr_stemmer21.default)(lunr2);
  (0, import_lunr42.default)(lunr2);
  (0, import_lunr41.default)(lunr2);
  return lunr2.vi;
}

// src/indexer/langs/ja.ts
var import_lunr_stemmer22 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr43 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr44 = __toESM(require("lunr-languages/lunr.ja"));
var import_tinyseg = __toESM(require("lunr-languages/tinyseg"));
function ja(lunr2) {
  (0, import_lunr_stemmer22.default)(lunr2);
  (0, import_lunr44.default)(lunr2);
  (0, import_tinyseg.default)(lunr2);
  (0, import_lunr43.default)(lunr2);
  return lunr2.ja;
}

// src/indexer/langs/jp.ts
var import_lunr_stemmer23 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr45 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr46 = __toESM(require("lunr-languages/lunr.jp"));
var import_tinyseg2 = __toESM(require("lunr-languages/tinyseg"));
function jp(lunr2) {
  (0, import_lunr_stemmer23.default)(lunr2);
  (0, import_lunr46.default)(lunr2);
  (0, import_tinyseg2.default)(lunr2);
  (0, import_lunr45.default)(lunr2);
  return lunr2.jp;
}

// src/indexer/langs/th.ts
var import_lunr_stemmer24 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr47 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr48 = __toESM(require("lunr-languages/lunr.th"));
var import_wordcut = __toESM(require("lunr-languages/wordcut"));
function th(lunr2) {
  (0, import_lunr_stemmer24.default)(lunr2);
  (0, import_lunr48.default)(lunr2);
  (0, import_wordcut.default)(lunr2);
  (0, import_lunr47.default)(lunr2);
  return lunr2.th;
}

// src/indexer/langs/hi.ts
var import_lunr_stemmer25 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr49 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr50 = __toESM(require("lunr-languages/lunr.hi"));
var import_wordcut2 = __toESM(require("lunr-languages/wordcut"));
function hi(lunr2) {
  (0, import_lunr_stemmer25.default)(lunr2);
  (0, import_lunr50.default)(lunr2);
  (0, import_wordcut2.default)(lunr2);
  (0, import_lunr49.default)(lunr2);
  return lunr2.hi;
}

// src/indexer/langs/ta.ts
var import_lunr_stemmer26 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr51 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr52 = __toESM(require("lunr-languages/lunr.ta"));
var import_wordcut3 = __toESM(require("lunr-languages/wordcut"));
function ta(lunr2) {
  (0, import_lunr_stemmer26.default)(lunr2);
  (0, import_lunr52.default)(lunr2);
  (0, import_wordcut3.default)(lunr2);
  (0, import_lunr51.default)(lunr2);
  return lunr2.ta;
}

// src/indexer/langs/sa.ts
var import_lunr_stemmer27 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr53 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr54 = __toESM(require("lunr-languages/lunr.sa"));
var import_wordcut4 = __toESM(require("lunr-languages/wordcut"));
function sa(lunr2) {
  (0, import_lunr_stemmer27.default)(lunr2);
  (0, import_lunr54.default)(lunr2);
  (0, import_wordcut4.default)(lunr2);
  (0, import_lunr53.default)(lunr2);
  return lunr2.sa;
}

// src/indexer/langs/kn.ts
var import_lunr_stemmer28 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr55 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr56 = __toESM(require("lunr-languages/lunr.kn"));
var import_wordcut5 = __toESM(require("lunr-languages/wordcut"));
function kn(lunr2) {
  (0, import_lunr_stemmer28.default)(lunr2);
  (0, import_lunr56.default)(lunr2);
  (0, import_wordcut5.default)(lunr2);
  (0, import_lunr55.default)(lunr2);
  return lunr2.kn;
}

// src/indexer/langs/te.ts
var import_lunr_stemmer29 = __toESM(require("lunr-languages/lunr.stemmer.support"));
var import_lunr57 = __toESM(require("lunr-languages/lunr.multi"));
var import_lunr58 = __toESM(require("lunr-languages/lunr.te"));
var import_wordcut6 = __toESM(require("lunr-languages/wordcut"));
function te(lunr2) {
  (0, import_lunr_stemmer29.default)(lunr2);
  (0, import_lunr58.default)(lunr2);
  (0, import_wordcut6.default)(lunr2);
  (0, import_lunr57.default)(lunr2);
  return lunr2.te;
}

// src/indexer/langs/index.ts
var langs = { ar, da, de, du, el, es, fi, fr, he, hu, hy, it, ko, nl, no, pt, ro, ru, sv, tr, vi, ja, jp, th, hi, ta, sa, kn, te };

// src/indexer/html.ts
var import_node_html_parser = require("node-html-parser");
function isElementNode(node) {
  return node.nodeType === import_node_html_parser.NodeType.ELEMENT_NODE;
}
function isTextNode(node) {
  return node.nodeType === import_node_html_parser.NodeType.TEXT_NODE;
}
function isBlockTag(tag) {
  return "div|p|li|td|section|br".includes(tag);
}
function html2text(html) {
  const node = (0, import_node_html_parser.parse)(html, {
    lowerCaseTagName: true,
    blockTextElements: {
      script: false,
      noscript: false,
      style: false,
      pre: true
    }
  });
  let currentBlock = [];
  let prependWhitespace = false;
  const blocks = [currentBlock];
  const work = [node];
  while (work.length) {
    const node2 = work.shift();
    if (!node2) {
      if (currentBlock.length) {
        blocks.push(currentBlock = []);
        prependWhitespace = false;
      }
      continue;
    }
    if (isElementNode(node2)) {
      const tagName = (node2.tagName || "").toLowerCase();
      if (node2.attributes["data-no-index"]) {
        continue;
      }
      if (tagName === "pre") {
        currentBlock.push(html2text(node2.rawText));
      } else if (isBlockTag(tagName)) {
        work.unshift(null, ...node2.childNodes, null);
      } else {
        work.unshift(...node2.childNodes);
      }
    }
    if (isTextNode(node2)) {
      if (node2.isWhitespace) {
        prependWhitespace = true;
      } else {
        let text = node2.text;
        if (prependWhitespace) {
          text = " " + text;
          prependWhitespace = false;
        }
        currentBlock.push(text);
      }
    }
  }
  return blocks.map(
    (block) => block.join(" ").trim().replace(/\s{2,}/g, " ").replace(/\s([.,!?:;])/g, "$1")
  ).join("\n").trimEnd();
}

// src/indexer/index.ts
var ReleaseFormat = /* @__PURE__ */ ((ReleaseFormat2) => {
  ReleaseFormat2["JSONP"] = "jsonp";
  ReleaseFormat2["RAW"] = "raw";
  return ReleaseFormat2;
})(ReleaseFormat || {});
var Indexer = class {
  constructor() {
    this.indices = {};
    this.docs = {};
  }
  /**
   * Stores list of current non empty language indexes.
   */
  get langs() {
    return Object.keys(this.indices);
  }
  /**
   * Adds new entry to search index
   *
   * @param lang - index language
   * @param url - document url, used as index uniq id
   * @param data - document json data
   *
   * @returns {void}
   */
  add(lang30, url, data) {
    if (!this.indices[lang30]) {
      this.init(lang30);
    }
    const { leading, title = "", meta = {} } = data;
    if (leading || meta.noindex || meta.noIndex) {
      return;
    }
    const content = html2text(data.html);
    const keywords = meta.keywords || [];
    this.docs[lang30][url] = { title, content, keywords };
    this.indices[lang30].add({
      ...this.docs[lang30][url],
      url
    });
  }
  /**
   * Dumps index and registry for target language.
   *
   * @param lang - index language
   * @param format - output format
   *
   * @returns {{index: Index, registry: Registry}}
   */
  release(lang30, format = "jsonp" /* JSONP */) {
    const index = this.indices[lang30].build();
    const registry = this.docs[lang30];
    if (format === "jsonp" /* JSONP */) {
      return {
        index: "self.index=" + JSON.stringify(index),
        registry: "self.registry=" + JSON.stringify(registry)
      };
    }
    return { index, registry };
  }
  init(lang30) {
    const index = new import_lunr59.Builder();
    if (langs[lang30]) {
      langs[lang30](import_lunr59.default);
      index.use(import_lunr59.default.multiLanguage("en", lang30));
    }
    index.ref("url");
    for (const [field, boost] of Object.entries(INDEX_FIELDS)) {
      index.field(field, { boost });
    }
    index.metadataWhitelist = ["position"];
    this.indices[lang30] = index;
    this.docs[lang30] = {};
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Indexer,
  ReleaseFormat,
  html2text
});
