// @diplodoc/algolia-extension/dist/config/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.options = void 0;
const config_1 = require("@diplodoc/cli/lib/config");
exports.options = {
    input: (0, config_1.option)({
        flags: '-i, --input <path>',
        desc: 'Path to documentation directory',
        required: true,
        default: './',
    }),
    appId: (0, config_1.option)({
        flags: '--app-id <id>',
        desc: 'Algolia Application ID',
        env: 'ALGOLIA_APP_ID',
    }),
    apiKey: (0, config_1.option)({
        flags: '--api-key <key>',
        desc: 'Algolia API Key',
        env: 'ALGOLIA_API_KEY',
    }),
    indexName: (0, config_1.option)({
        flags: '--index-name <name>',
        desc: 'Algolia Index Name',
        env: 'ALGOLIA_INDEX_NAME',
    }),
    index: (0, config_1.option)({
        flags: '--index [boolean]',
        desc: 'Whether to create and upload an index for search',
    }),
    searchApiKey: (0, config_1.option)({
        flags: '--search-api-key <key>',
        desc: 'Client-side API key for search',
        env: 'ALGOLIA_SEARCH_API_KEY',
    }),
    provider: (0, config_1.option)({
        flags: '--search-provider <name>',
        desc: 'Search provider name',
        env: 'ALGOLIA_PROVIDER',
        default: 'algolia',
    }),
    api: (0, config_1.option)({
        flags: '--search-api <path>',
        desc: 'Path to the client-side search API',
        env: 'ALGOLIA_API_PATH',
        default: '_search/api.js',
    }),
};
