// @diplodoc/algolia-extension/dist/workers/processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const worker_threads_1 = require("worker_threads");
const document_processor_1 = require("../core/document-processor");
if (!worker_threads_1.parentPort) {
    throw new Error('This file should be run as a worker thread');
}
function sendResult(records) {
    const resultMessage = {
        type: 'result',
        data: { records },
    };
    if (worker_threads_1.parentPort) {
        worker_threads_1.parentPort.postMessage(resultMessage);
    }
}
function sendError(error) {
    const errorMessage = {
        type: 'error',
        data: {
            message: error instanceof Error ? error.message : String(error),
            stack: error instanceof Error ? error.stack : undefined,
        },
    };
    if (worker_threads_1.parentPort) {
        worker_threads_1.parentPort.postMessage(errorMessage);
    }
}
process.on('uncaughtException', (error) => {
    sendError(error);
});
process.on('unhandledRejection', (reason) => {
    sendError(new Error(`Unhandled promise rejection: ${reason}`));
});
worker_threads_1.parentPort.on('message', (message) => {
    try {
        if (message.type === 'terminate') {
            process.exit(0);
            return;
        }
        if (message.type === 'process') {
            const processMessage = message;
            const { path, lang, html, title, meta, skipHtmlExtension } = processMessage.data;
            if (meta.noIndex) {
                sendResult([]);
                return;
            }
            const records = (0, document_processor_1.processDocument)({ path, lang, html, title, meta, skipHtmlExtension });
            sendResult(records);
        }
    }
    catch (error) {
        sendError(error);
    }
});
